pkg/trace/api/api.go

```diff
@@ -210,9 +212,19 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 
 // handleTraces knows how to handle a bunch of traces
 func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.Request) {
+	if atomic.LoadInt64(&r.refuse) == 1 {
+		// using too much memory
+		io.Copy(ioutil.Discard, req.Body)
+		w.WriteHeader(http.StatusNotAcceptable)
+		io.WriteString(w, "request rejected; trace-agent is past memory threshold (apm_config.max_memory)")
+		metrics.Gauge("datadog.trace_agent.receiver.refused", 1, []string{"reason:mem"}, 1)
+		return
+	}
 	if !r.PreSampler.Sample(req) {
```

Comment:
Instead of running this presampler, shouldn't we do the same here? Just stop accepting requests when we go over the threshold?

Reply:
Technically you could, the idea is that if you do so, you need to appropriately update the sample rate depending on wether you drop/keep the trace, which is basically what the presampler is responsible for. At some point, it sounds to me that keeping this rate up-to-date is something we want to encapsulate into some sort of independent module. So you'd end up rewriting the presampler. What I'm suggesting also here is that before, when we were dropping traces because the channel was full, that rate might have been wrong. Thinking of it, the logic might be, now, to:

* drop the trace if we're over CPU
* drop the trace if we're over Memory usage
* maintain the presampling rate accordingly

Reply:
Thanks for explaining. I think this is just an intermediate step to offer a build where the customer has a way out of losing traces. Before this change this wasn't possible. My question above was more like a curiosity.

Reply:
> Instead of running this presampler, shouldn't we do the same here? Just stop accepting requests when we go over the threshold?

Agreed, that would be easier to reason about. Ideally the agent only accepts a trace if it's actually going to send it.

Reply:
No, this is perfect. The presampler has a smoother algorithm in deciding how to rate limit. As an example, if CPU usage goes over the set threshold by 20%, the presampler will accept only 80% of requests, balancing out the CPU usage without spikes. If we were to complete stop dropping everything for the next 20 seconds, we'd be wasting a lot of good CPU time and losing a lot of traces.

In fact, the rate limiting based on memory above should be using the same logic as we do here, with the pre sampler. This will be follow up work after this PR.

At the time I made this comment, I did not have the above understanding.